### PR TITLE
Ajd 331 optimization 2nd stage

### PR DESCRIPTION
**Pull request by danielm1405**
## Types of PR

- [ ] New Features
- [x] Upgrade of existing features
- [ ] Bugfix

## Description

This PR aims to limit the recalculation of the trajectories. Each action has access to the reference trajectory that is precalculated on `traffic_simulator` level and passed to the behavior tree nodes. To determine entity actions, distance from the trajectory to other objects (entities, stop lines, traffic lights, etc.) is calculated multiple times. Only a part of the trajectory is taken into the consideration during these calculations - this part will be called trimmed trajectory.

Previously, the trimmed trajectory was obtained in the following way:
* waypoints were calculated based on the reference trajectory and the horizon (2.)
* the trajectory was calculated as a spline using the waypoints as knots (3.)
Moreover, the trajectory was recalculated multiple times in `tick()` as distance calculation functions required waypoints as an argument. 

This PR changes this flow. Instead of recalculating the trajectory, the trajectory is trimmed to match the desired starting and ending points (using `getTrimmed` method). This is done only once in each `tick()` call.

optimization-2nd-stage drawio

To sum up the advantages:
* previously: a **costly** spline calculation was performed **multiple** times in each `tick()` call
* now: relatively **cheap** spline trimming is performed **once** in each `tick()` call

### CODE REVIEW

File: simulation/behavior_tree_plugin/src/vehicle/follow_lane_sequence/stop_at_traffic_light_action.cpp [hunk 1]
```diff
@@ -102,13 +105,12 @@ BT::NodeStatus StopAtTrafficLightAction::tick()
   if (waypoints.waypoints.empty()) {
     return BT::NodeStatus::FAILURE;
   }
-  const auto spline = traffic_simulator::math::CatmullRomSpline(waypoints.waypoints);
   const auto distance_to_traffic_stop_line =
-    hdmap_utils->getDistanceToTrafficLightStopLine(route_lanelets, waypoints.waypoints);
+    hdmap_utils->getDistanceToTrafficLightStopLine(route_lanelets, *trajectory);
```
Comment (hakuturu583) [inline simulation/behavior_tree_plugin/src/vehicle/follow_lane_sequence/stop_at_traffic_light_action.cpp:109]: Currently, *trajectory cannot be null pointer.
However, please do null check for future development.

File: simulation/behavior_tree_plugin/src/vehicle/follow_lane_sequence/stop_at_traffic_light_action.cpp [hunk 2]
```diff
@@ -102,13 +105,12 @@ BT::NodeStatus StopAtTrafficLightAction::tick()
...
   if (!distance_to_traffic_stop_line) {
     return BT::NodeStatus::FAILURE;
   }
-  distance_to_stop_target_ = getDistanceToTrafficLightStopLine(route_lanelets, waypoints.waypoints);
+  distance_to_stop_target_ = getDistanceToTrafficLightStopLine(route_lanelets, *trajectory);
```
Comment (hakuturu583) [inline simulation/behavior_tree_plugin/src/vehicle/follow_lane_sequence/stop_at_traffic_light_action.cpp:113]: Currently, *trajectory cannot be null pointer.
However, please do null check for future development.

File: simulation/traffic_simulator/include/traffic_simulator/math/hermite_curve.hpp [hunk 3]
```diff
@@ -56,6 +56,7 @@ class HermiteCurve
   double getMaximum2DCurvature() const;
   double getLength(size_t num_points) const;
   double getLength() const { return length_; }
+  void setLength(double length) { length_ = length; }
```
Comment (hakuturu583) [inline simulation/traffic_simulator/include/traffic_simulator/math/hermite_curve.hpp:59]: Length should be calculated in hermite curve class.
Why do you add setLength function?

File: simulation/traffic_simulator/src/math/catmull_rom_spline.cpp [hunk 4]
```diff
@@ -228,9 +256,30 @@ CatmullRomSpline::CatmullRomSpline(const std::vector<geometry_msgs::msg::Point>
...
+{
+  curves_ = curves;
+
+  for (const auto & curve : curves_) {
+    length_list_.emplace_back(curve.getLength());
+    maximum_2d_curvatures_.emplace_back(curve.getMaximum2DCurvature());
+  }
+
+  length_list_.front() = staring_curve_length;
+  curves_.front().setLength(staring_curve_length);
```
Comment (hakuturu583) [inline simulation/traffic_simulator/src/math/catmull_rom_spline.cpp:274]: Why do you set length?

File: simulation/traffic_simulator/src/math/catmull_rom_spline.cpp [hunk 5]
```diff
@@ -228,9 +256,30 @@ CatmullRomSpline::CatmullRomSpline(const std::vector<geometry_msgs::msg::Point>
...
+  length_list_.back() = ending_curve_length;
+  curves_.back().setLength(ending_curve_length);
```
Comment (hakuturu583) [inline simulation/traffic_simulator/src/math/catmull_rom_spline.cpp:276]: Why do you set length?

**Review by hakuturu583 — APPROVED** [review]
This approval was my mistake.
Sorry.
Please ignore this approval.

**Review by hakuturu583 — CHANGES_REQUESTED** [review]
Also, please update Release Note

**Comment by hakuturu583** [comment]
@danielm1405 It seems, this feature is huge change, so please write description more detail.
Please refer [this PR](https://github.com/tier4/scenario_simulator_v2/pull/729).

File: simulation/traffic_simulator/include/traffic_simulator/math/hermite_curve.hpp (hunk 3 above)
Comment (danielm1405) [reply simulation/traffic_simulator/include/traffic_simulator/math/hermite_curve.hpp:59]: This is needed for curve trimming. This PR aims to avoid spline recalculation because this operation is very costly. Instead, the `trajectory` is calculated as a section of `reference_trajectory`. To calculate the subspline (which is just a part of the spline) you need to determine which curves should be included in the subspline. The middle curves are taken as a whole, but the first one and the last one need to be trimmed (because the start point of the subspline can be in the middle of the curve and we need only a part of it to construct the subspline). The curves should not change their shapes, but only change their length. So `setLength` is used to shorten the curves without the need to recalculate them from scratch (because it is costly as well).

File: simulation/traffic_simulator/src/math/catmull_rom_spline.cpp (hunk 4 above)
Comment (danielm1405) [reply simulation/traffic_simulator/src/math/catmull_rom_spline.cpp:274]: This is the trimming of the starting curve described [here](https://github.com/tier4/scenario_simulator_v2/pull/710#discussion_r842661253).

File: simulation/traffic_simulator/src/math/catmull_rom_spline.cpp (hunk 5 above)
Comment (danielm1405) [reply simulation/traffic_simulator/src/math/catmull_rom_spline.cpp:276]: This is the trimming of the ending curve described [here](https://github.com/tier4/scenario_simulator_v2/pull/710#discussion_r842661253).

File: simulation/traffic_simulator/include/traffic_simulator/math/hermite_curve.hpp (hunk 3 above)
Comment (hakuturu583) [reply simulation/traffic_simulator/include/traffic_simulator/math/hermite_curve.hpp:59]: I see, but I think this function is dangerous because of this is a public function.
And also, this function does not changes starting point and ending point of hermite curve it breaks the hermite curve class.

File: simulation/traffic_simulator/src/math/catmull_rom_spline.cpp (hunk 4 above)
Comment (hakuturu583) [reply simulation/traffic_simulator/src/math/catmull_rom_spline.cpp:274]: I see, but I think this function is dangerous because of this is a public function.
And also, this function does not changes starting point and ending point of hermite curve it breaks the hermite curve class.

File: simulation/traffic_simulator/src/math/catmull_rom_spline.cpp (hunk 5 above)
Comment (hakuturu583) [reply simulation/traffic_simulator/src/math/catmull_rom_spline.cpp:276]: I see, but I think this function is dangerous because of this is a public function.
And also, this function does not changes starting point and ending point of hermite curve it breaks the hermite curve class.

**Review by hakuturu583 — CHANGES_REQUESTED** [review]

File: simulation/traffic_simulator/include/traffic_simulator/math/hermite_curve.hpp (hunk 3 above)
Comment (danielm1405) [reply simulation/traffic_simulator/include/traffic_simulator/math/hermite_curve.hpp:59]: Yes, I agree this solution is not perfect. I will try to come up with a better idea. If you have any suggestions on how to implement hermite curve trimming I would really appreciate it.

Comment (hakuturu583) [reply simulation/traffic_simulator/include/traffic_simulator/math/hermite_curve.hpp:59]: I think problem is setting length.
Can you use constructor of HermiteCurve class?

Comment (danielm1405) [reply simulation/traffic_simulator/include/traffic_simulator/math/hermite_curve.hpp:59]: We discussed this problem internally. We agree that length setting is a problem. We came up with a solution: we will introduce the `CatmullRomSubspline` class. It stores a reference to the original `CatmullRomSpline`, `start_s` and `end_s`. It provides an interface for accessing the underlying spline as it was trimmed (as it would start in `start_s` and end in `end_s`). The underlying spline would not be changed in any way. Therefore, the dangerous functions `HermiteCurve::setLength` and `CatmullRomSpline::getSubspline` will no longer be needed and will be removed.

What do you think about this approach?

Comment (hakuturu583) [reply simulation/traffic_simulator/include/traffic_simulator/math/hermite_curve.hpp]: Do you mean you trimmed spline in CatmullRomSubspline class not in HermiteCurve class?
And s value of the Hermite curve class is always from 0 to it's length?
If so, I think it is a good idea!

Comment (danielm1405) [reply simulation/traffic_simulator/include/traffic_simulator/math/hermite_curve.hpp]: I believe it's even better - I do not trim the spline at all. I use the `CatmullRomSubspline` class to access the spline as it was trimmed, but the underlying spline is not modified at all.

**Comment by hakuturu583** [comment]
@danielm1405 Please tell me why you need CatmullRomInterface class?

File: simulation/traffic_simulator/include/traffic_simulator/math/catmull_rom_interface.hpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+#include <geometry_msgs/msg/point.hpp>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace traffic_simulator
+{
+namespace math
+{
+class CatmullRomInterface
```
Comment (hakuturu583) [inline simulation/traffic_simulator/include/traffic_simulator/math/catmull_rom_interface.hpp:29]: I think it's naming is too ambiguous.
I think it is a base class for catmull rom spline curve.
How about rename to CatmullRomSplineBase or CatmullRomSplineInterface class.

**Review by hakuturu583 — CHANGES_REQUESTED** [review]

File: simulation/traffic_simulator/include/traffic_simulator/math/catmull_rom_interface.hpp (hunk 6 above)
Comment (danielm1405) [reply simulation/traffic_simulator/include/traffic_simulator/math/catmull_rom_interface.hpp]: Ok, I renamed `CatmullRomInterface` to `CatmullRomSplineInterface`

**Review by hakuturu583 — APPROVED** [review]
